File: leetcode/1302_Deepest_Leaves_Sum.py
from typing import Optional, Tuple, List
from unittest import TestCase

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def deepestLeavesSum(self, root: Optional[TreeNode]) -> int:
        max_level, max_level_sum = 0, 0
        if root is not None:
            stack = [(0, root)]
            while stack:
                level, v = stack.pop()
                if max_level < level:
                    max_level, max_level_sum = level, v.val
                elif level == max_level:
                    max_level_sum += v.val
                if v.left is not None:
                    stack.append((level + 1, v.left))
                if v.right is not None:
                    stack.append((level + 1, v.right))
        return max_level_sum

class DeepestLeavesSumTest(TestCase):
    def test_1(self):
        self.assertEqual(15, Solution().deepestLeavesSum(TreeNode(1, TreeNode(2, TreeNode(4, TreeNode(7)), TreeNode(5)), TreeNode(3, None, TreeNode(6, None, TreeNode(8))))))

            

File: leetcode/1302_Deepest_Leaves_Sum.rs
struct Solution {}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None
        }
    }
}

use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn deepest_leaves_sum(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let (mut max_level, mut max_level_value) = (0, 0);
        if let Some(node) = root {
            let mut stack = vec![(0, node)];
            while let Some((level, node)) = stack.pop() {
                if level > max_level {
                    max_level = level;
                    max_level_value = node.borrow().val;
                } else if level == max_level {
                    max_level_value += node.borrow().val;
                }
                if let Some(child) = node.borrow().left.as_ref() {
                    stack.push((level + 1, child.clone()));
                }
                if let Some(child) = node.borrow().right.as_ref() {
                    stack.push((level + 1, child.clone()));
                }
            }
        }
        max_level_value
    }
}